File: iosApp/iosApp/Recurring/ViewModels/RecurringFormViewModel.swift
import Foundation

@Observable
final class RecurringFormViewModel {
    enum ContentState {
        case loading
        case ready
        case error(String)
    }

    private let bridge: any RecurringRepositoryBridge

    var contentState: ContentState = .ready
    var formState = RecurringFormState()
    var showDeleteConfirmation = false
    var didSaveSuccessfully = false
    var toastMessage: String?
    var isEditMode: Bool { formState.isEditMode }

    init(
        bridge: any RecurringRepositoryBridge = AppDependencies.shared.recurringBridge
    ) {
        self.bridge = bridge
    }

    func loadTemplate(id: String) async {
        contentState = .loading
        do {
            guard let template = try await bridge.loadTemplateById(id: id) else {
                contentState = .error("Template not found")
                return
            }
            formState.populate(from: template)
            contentState = .ready
        } catch {
            contentState = .error(error.localizedDescription)
        }
    }

    func save() async {
        guard formState.validate() else { return }
        formState.isSaving = true

        do {
            let amount = Double(formState.amountText) ?? 0

            if let editId = formState.templateId {
                _ = try await bridge.updateTemplate(
                    id: editId,
                    amount: amount,
                    type: formState.selectedType,
                    category: formState.selectedCategory,
                    note: formState.noteText,
                    frequency: formState.selectedFrequency,
                    startDateMillis: formState.startDateMillis,
                    endDateMillis: formState.endDateMillis
                )
                toastMessage = String(localized: "recurring_updated_snackbar")
            } else {
                _ = try await bridge.createTemplate(
                    amount: amount,
                    type: formState.selectedType,
                    category: formState.selectedCategory,
                    note: formState.noteText,
                    frequency: formState.selectedFrequency,
                    startDateMillis: formState.startDateMillis,
                    endDateMillis: formState.endDateMillis
                )
                toastMessage = String(localized: "recurring_created_snackbar")
            }

            formState.isSaving = false
            didSaveSuccessfully = true
        } catch {
            formState.isSaving = false
            toastMessage = error.localizedDescription
        }
    }

    func delete() async {
        guard let id = formState.templateId else { return }
        showDeleteConfirmation = false

        do {
            try await bridge.deleteTemplate(id: id)
            toastMessage = String(localized: "recurring_deleted_snackbar")
            didSaveSuccessfully = true
        } catch {
            toastMessage = error.localizedDescription
        }
    }
}

File: feature/recurring-transactions/impl/src/commonMain/kotlin/com/expense/tracker/feature/recurring/impl/RecurringFormViewModel.kt
package com.expense.tracker.feature.recurring.impl

import com.expense.tracker.feature.expense.domain.model.ExpenseCategory
import com.expense.tracker.feature.expense.domain.model.IncomeCategory
import com.expense.tracker.feature.expense.domain.model.TransactionType
import com.expense.tracker.feature.recurring.domain.repository.RecurringTemplateRepository
import com.expense.tracker.shared.core.domain.AppError
import com.expense.tracker.shared.core.domain.Result
import com.expense.tracker.shared.core.domain.TimeProvider
import com.expense.tracker.shared.core.domain.asMessageText
import com.expense.tracker.shared.core.presentation.MviViewModel

class RecurringFormViewModel(
    private val repository: RecurringTemplateRepository,
    private val timeProvider: TimeProvider,
) : MviViewModel<RecurringFormState, RecurringFormAction, RecurringFormEvent>(
    initialState = RecurringFormState(
        startDateMillis = timeProvider.nowMillis(),
    ),
) {
    override suspend fun handleAction(action: RecurringFormAction) {
        when (action) {
            is RecurringFormAction.SetTemplate -> setTemplate(action.templateId)
            is RecurringFormAction.AmountChanged -> updateState { it.copy(amountText = action.value) }
            is RecurringFormAction.TypeSelected -> updateState {
                val firstCategory = when (action.type) {
                    TransactionType.INCOME -> IncomeCategory.entries.first().name
                    TransactionType.EXPENSE -> ExpenseCategory.entries.first().name
                }
                it.copy(selectedType = action.type, selectedCategory = firstCategory)
            }
            is RecurringFormAction.CategorySelected -> updateState {
                it.copy(selectedCategory = action.category, categoryMenuExpanded = false)
            }
            is RecurringFormAction.ToggleCategoryMenu -> updateState {
                it.copy(categoryMenuExpanded = !it.categoryMenuExpanded)
            }
            is RecurringFormAction.DismissCategoryMenu -> updateState {
                it.copy(categoryMenuExpanded = false)
            }
            is RecurringFormAction.NoteChanged -> updateState { it.copy(noteText = action.value) }
            is RecurringFormAction.FrequencySelected -> updateState {
                it.copy(selectedFrequency = action.frequency)
            }
            is RecurringFormAction.StartDateSelected -> updateState {
                it.copy(startDateMillis = action.millis)
            }
            is RecurringFormAction.EndDateSelected -> updateState {
                it.copy(endDateMillis = action.millis)
            }
            is RecurringFormAction.ClearEndDate -> updateState {
                it.copy(endDateMillis = null)
            }
            is RecurringFormAction.Save -> save()
        }
    }

    private suspend fun setTemplate(templateId: String?) {
        updateState { it.copy(templateId = templateId) }
        if (templateId != null) {
            loadTemplate(templateId)
        }
    }

    private suspend fun loadTemplate(id: String) {
        updateState { it.copy(isLoading = true) }

        when (val result = repository.loadTemplateById(id)) {
            is Result.Success -> {
                val template = result.value
                if (template != null) {
                    updateState {
                        it.copy(
                            isLoading = false,
                            amountText = template.amount.toString(),
                            selectedType = template.type,
                            selectedCategory = template.category,
                            noteText = template.note,
                            selectedFrequency = template.frequency,
                            startDateMillis = template.startDateMillis,
                            endDateMillis = template.endDateMillis,
                        )
                    }
                } else {
                    updateState { it.copy(isLoading = false, error = "Template not found") }
                }
            }

            is Result.Failure -> updateState {
                it.copy(isLoading = false, error = result.error.asMessageText())
            }
        }
    }

    private suspend fun save() {
        val currentState = state.value
        val amount = currentState.amountText.toDoubleOrNull()
        val isAmountValid = amount != null && amount > 0
        val startDate = currentState.startDateMillis

        if (!isAmountValid || startDate == null) {
            if (!isAmountValid) {
                sendEvent(RecurringFormEvent.Error("Please enter a valid amount"))
            } else {
                sendEvent(RecurringFormEvent.Error("Please select a start date"))
            }
            return
        }

        updateState { it.copy(isSaving = true) }

        val result = if (currentState.templateId != null) {
            repository.updateTemplate(
                id = currentState.templateId,
                amount = amount,
                type = currentState.selectedType,
                category = currentState.selectedCategory,
                note = currentState.noteText.trim(),
                frequency = currentState.selectedFrequency,
                startDateMillis = startDate,
                endDateMillis = currentState.endDateMillis,
            )
        } else {
            repository.createTemplate(
                amount = amount,
                type = currentState.selectedType,
                category = currentState.selectedCategory,
                note = currentState.noteText.trim(),
                frequency = currentState.selectedFrequency,
                startDateMillis = startDate,
                endDateMillis = currentState.endDateMillis,
            )
        }

        when (result) {
            is Result.Success -> {
                updateState { it.copy(isSaving = false) }
                sendEvent(RecurringFormEvent.RecurringSaved)
            }

            is Result.Failure -> {
                updateState { it.copy(isSaving = false) }
                sendEvent(RecurringFormEvent.Error(result.error.asMessageText()))
            }
        }
    }
}
